tests/linearizability/model.go

```diff
@@ -99,22 +98,32 @@ func step(state EtcdState, request etcdRequest, response etcdResponse) (bool, Et
 	switch request.op {
 	case Get:
 		if state.Value == response.getData {
-			return true, state
+			if state.LastRevision == 0 {
+				state.LastRevision = response.revision
+			}
+			if state.LastRevision == response.revision {
+				return true, state
+			}
 		}
-		for write := range state.FailedWrites {
-			if write == response.getData {
+		if state.LastRevision < response.revision {
+			_, ok := state.FailedWrites[response.getData]
+			if ok {
 				state.Value = response.getData
-				delete(state.FailedWrites, write)
+				state.LastRevision = response.revision
+				delete(state.FailedWrites, response.getData)
 				return true, state
 			}
 		}
 	case Put:
-		if response.err == nil {
-			state.Value = request.putData
-		} else {
+		if response.err != nil {
 			state.FailedWrites[request.putData] = struct{}{}
+			return true, state
+		}
+		if state.LastRevision < response.revision {
+			state.Value = request.putData
+			state.LastRevision = response.revision
+			return true, state
 		}
```

Comment:
My previous comment https://github.com/etcd-io/etcd/pull/14685/files#r1014471776 might still valid.

Since `response.err` is `nill`, so we should save the `request.putData` and `response.revision` anyway. 

```suggestion
		lastRev := state.LastRevision
		state.Value = request.putData
		state.LastRevision = response.revision
		if lastRev < response.revision {
			return true, state
		}
```

Reply:
Commit suggestion you proposed doesn't change the logic at all, you just save data before you validate the state transition is correct. 

> We might need to update state.LastRevision as long as response.revision != 0

Please remember that revision is a positive number, so checking that revision is increasing like in `state.LastRevision < response.revision` already ensures that `response.revision != 0`

Reply:
As long as `response.err` is nil, then we need to save `request.putData` and `response.revision` to state. While your code only save them when `state.LastRevision < response.revision` and of course `response.err `is nil,.

Reply:
> As long as response.err is nil, then we need to save request.putData and response.revision to state

That's not true. Model should only allow operations that would be correct for etcd. Etcd revision should never decrease. That's why I put `state.LastRevision < response.revision` check.